File: frontend/src/types/aiTelemetry.ts
// ── AI Telemetry types ──────────────────────────────────────────────────────

export interface AiTelemetryStats {
  total_calls: number;
  successful_calls: number;
  failed_calls: number;
  total_tokens: number;
  total_prompt_tokens: number;
  total_completion_tokens: number;
  total_cost_usd: number;
  avg_latency_ms: number;
  avg_tokens_per_call: number;
}

export interface AiTelemetryStatsResponse {
  stats: AiTelemetryStats;
}

export interface DailyBreakdownPoint {
  date: string;
  calls: number;
  tokens: number;
  cost_usd: number;
  avg_latency_ms: number;
}

export interface DailyBreakdownResponse {
  points: DailyBreakdownPoint[];
}

export interface PerEndpointUsage {
  endpoint_id: string;
  endpoint_name: string;
  total_calls: number;
  total_tokens: number;
  total_cost_usd: number;
  avg_latency_ms: number;
  success_rate: number;
}

export interface PerEndpointUsageResponse {
  endpoints: PerEndpointUsage[];
}

export interface AiHealthResponse {
  total_calls: number;
  success_rate: number;
  error_rate: number;
  avg_latency_ms: number;
  last_error: string | null;
  last_error_at: string | null;
  model_name: string;
}

File: app/schemas/ai_telemetry.py
"""AI Telemetry schemas — response models for the telemetry dashboard API."""

import uuid
from datetime import datetime

from pydantic import BaseModel


# ── Stats aggregate ──────────────────────────────────────────────────────────


class AiTelemetryStats(BaseModel):
    """Aggregate KPIs across all AI calls for the tenant."""

    total_calls: int
    successful_calls: int
    failed_calls: int
    total_tokens: int
    total_prompt_tokens: int
    total_completion_tokens: int
    total_cost_usd: float
    avg_latency_ms: float
    avg_tokens_per_call: float


class AiTelemetryStatsResponse(BaseModel):
    stats: AiTelemetryStats


# ── Daily breakdown ──────────────────────────────────────────────────────────


class DailyBreakdownPoint(BaseModel):
    """One day of AI usage — for the daily time-series chart."""

    date: str  # ISO date, e.g. "2025-02-20"
    calls: int
    tokens: int
    cost_usd: float
    avg_latency_ms: float


class DailyBreakdownResponse(BaseModel):
    points: list[DailyBreakdownPoint]


# ── Per-endpoint usage ───────────────────────────────────────────────────────


class PerEndpointUsage(BaseModel):
    """AI usage ranked by endpoint — for the cost-by-endpoint chart."""

    endpoint_id: uuid.UUID
    endpoint_name: str
    total_calls: int
    total_tokens: int
    total_cost_usd: float
    avg_latency_ms: float
    success_rate: float


class PerEndpointUsageResponse(BaseModel):
    endpoints: list[PerEndpointUsage]


# ── Health ───────────────────────────────────────────────────────────────────


class AiHealthResponse(BaseModel):
    """LLM health metrics — success rate, error rate, recent errors."""

    total_calls: int
    success_rate: float
    error_rate: float
    avg_latency_ms: float
    last_error: str | None
    last_error_at: datetime | None
    model_name: str
